### ISSUE TITLE
Renovate custom managers: auto-bump Terraform / AWS CLI versions + dependency-coverage audit

### ISSUE THREAD

**Issue by bgauduch** [bgauduch/terraform-aws-cli#20]
> **Original ask (2020-05-04, preserved verbatim):** _"using github action, maybe open a new PR with updated version?"_
>
> Qualified below per the intake process (ADR-0014). This issue is the live home for the deferred Renovate **custom-manager** work.
>
> **⚠️ Reconciliation (2026-07-28) with harness study #152:** #152 arbitrates this issue — **Renovate owns _detect_** (customManager bumps `supported_versions.json`), a **script owns _materialise_** (`scripts/bump-version.sh`: fetch `SHA256SUMS` + `.sig`, GPG-verify, drop the superseded patch per ADR-0004, sync the test template — Renovate cannot do this). So #20 is the **trigger**, not the whole job; its **realisation is folded into #152 PR 4** (`feat(scripts): version and pin automation`). This **resolves the GPG open question** below (→ automated by the script, not manual). Sequencing: **#151 (ADR-0014) merges first**, then #152 go/nogo.
>
> **2026-09-20:** App installed (Dependency Dashboard live), required status checks configured on `master`, automerge decided — minor/patch/pin/digest merge on green, majors human-merged (ADR-0012, amended; the delegation and the `adr-not-needed` labelling deliver via the amendment's PR). The runner version (`ubuntu-22.04`) joins Task 4's batch. The materialise half exists (`bump-version.sh`, ADR-0021). **Remaining scope here: Task 1, the detect custom managers.**

## Context

Native Renovate config shipped in **#122** (config:recommended, grouped GitHub Actions + Debian base image, automerge patch/pin/digest, `chore(deps)`). The **custom manager** for the tool versions in `supported_versions.json` was **deferred there** ("needs datasource design"); epic **#102** was then closed as completed, re-homing the sub-task **here**; roadmap **Phase 1** still carries the line. Decision baseline: **ADR-0002** (Renovate only). Process: **ADR-0014**. Automation arbitration: **#152**.

## Dependency-coverage audit (2026-07-24)

Goal: verify every repo dependency is covered by Renovate, natively or custom.

| Dependency | Location | Manager | Covered? |
|---|---|---|---|
| GitHub Actions (checkout, setup-qemu/buildx, build-push, release-please, hadolint-action, setup-node, dockerhub-description, container-structure-test-action) | `.github/workflows/**` | `github-actions` (native) | ✅ yes, grouped |
| Debian base image (sha256 digest) | `Dockerfile` `ARG DEBIAN_VERSION` | `dockerfile` (native) | ✅ yes, grouped |
| **Terraform versions** (`tf_versions[]`) | `supported_versions.json` | — | ❌ **no — this issue (detect); materialise = `bump-version.sh` (ADR-0021)** |
| **AWS CLI versions** (`awscli_versions[]`) | `supported_versions.json` | — | ❌ **no — this issue (detect); materialise = `bump-version.sh` (ADR-0021)** |
| APT pinned packages (ca-certificates, curl, gnupg, unzip, git, jq, openssh-client) | `Dockerfile` (ADR-0010) | — | ❌ no — see Task 2 (`refresh-apt-pins.sh` route, ADR-0021) |
| Tool image `hadolint/hadolint` | `scripts/validate.sh` | — | ❌ no — see Task 3 |
| Tool image `container-structure-test` | `scripts/validate.sh` | — | ❌ no — see Task 3 |
| Action digest pinning (SHA) | workflows `@vX` tags | — | ❌ not configured — see Task 4 (→ #152 PR 6 / #99) |
| commitlint npm packages | `commitlint.yml` (`npm install`, unpinned) | — | ⚪ n/a (nothing versioned to track) |

**Verdict:** not all dependencies are covered. Native managers handle Actions + base image; the tool versions (this issue) and three adjacent gaps remain.

## Salvage: the #109 draft custom managers

Closed PR **#109** already drafted working custom regex managers. Starting point:

```json
"customManagers": [
  {
    "description": "Track latest Terraform version in supported_versions.json",
    "customType": "regex",
    "fileMatch": ["^supported_versions\\.json$"],
    "matchStrings": ["\"tf_versions\":\\s*\\[[^\\]]*\"(?<currentValue>[0-9]+\\.[0-9]+\\.[0-9]+)\"\\s*\\]"],
    "depNameTemplate": "hashicorp/terraform",
    "datasourceTemplate": "github-releases",
    "extractVersionTemplate": "^v(?<version>.*)$",
    "versioningTemplate": "semver"
  },
  {
    "description": "Track latest AWS CLI version in supported_versions.json",
    "customType": "regex",
    "fileMatch": ["^supported_versions\\.json$"],
    "matchStrings": ["\"awscli_versions\":\\s*\\[[^\\]]*\"(?<currentValue>[0-9]+\\.[0-9]+\\.[0-9]+)\"\\s*\\]"],
    "depNameTemplate": "aws/aws-cli",
    "datasourceTemplate": "github-tags",
    "versioningTemplate": "semver"
  }
]
```

**Caveats to resolve before reusing (from a review of #109):**
1. **Latest-only capture.** The regex anchors on the closing `]`, so it only matches the **last** element of each array — it will not bump intermediate versions nor introduce a new minor. Confirm this matches intent (the matrix keeps one patch per minor; only the newest slot auto-bumps).
2. **Pre-release filtering.** Terraform GitHub releases include `-alpha/-beta/-rc`; ensure unstable is excluded (Renovate `ignoreUnstable` default should cover it — verify).
3. **Scope of #109 is incomplete.** It covered neither APT pins, nor the tool images, nor action digest pinning (Tasks 2–4 below).

## Key design constraint (why this was deferred)

A Renovate version bump alone **does not yield a buildable image**: each new Terraform/AWS CLI version needs its **GPG verification files under `security/`** (`.asc`, `SHA256SUMS`, `.sig`) — see `docs/binaries-verifications.md`. **Resolved:** `scripts/bump-version.sh` (ADR-0021) materialises `security/` deterministically after Renovate bumps the JSON. The earlier "reminder-only vs automate" question is therefore settled → automate via the script.

## Proposed scope & plan (intake phase 4 — go/nogo is GO)

**In scope for this issue (Task 1 — the _detect_ half):** Terraform + AWS CLI custom regex managers.
- [ ] Salvage #109's `customManagers` + version grouping `packageRules`.
- [ ] Resolve caveats 1–2; validate with `npx --package renovate renovate-config-validator`.
- [ ] **No automerge** on version PRs (they are invalid until `bump-version.sh` runs).
- [ ] Decide schedule (`every weekend` like #109, or immediate).
- [ ] Structural path (`renovate.json`) → ship with an **ADR** (custom-manager / datasource design) or amend **ADR-0002**. Decide at planning.

**Out of scope — follow-up issues (dedup/prioritise per ADR-0014):**
- **Task 2 — APT pin custom manager** (`deb` datasource) or the deterministic `refresh-apt-pins.sh` route (ADR-0021). Today run on demand (`docs/dependencies-upgrades.md`, ADR-0010).
- **Task 3 — oracle tool images** (`hadolint`, `container-structure-test` pins in `scripts/validate.sh`): custom regex manager, or relocate the pins so the native managers see them. Note the action variants are already covered → drift risk.
- **Task 4 — Action digest pinning** (`helpers:pinGitHubActionDigests`): unblocks the `commitlint.yml` TODO; → #152 PR 6 / #99 (Phase 5). The runner version bump rides the same batch.

## Open question — trigger ("à voir")

Trigger/schedule for the Renovate detect PRs (immediate vs `every weekend`) — decide at Task 1 planning.

## Traceability

#20 (2020 seed) → epic **#102** (qualified) → **ADR-0002** (Renovate only) → roadmap **Phase 1** → **#122** (native slice shipped, custom manager deferred) → re-homed here → **arbitrated by #152** (detect vs materialise) → materialise shipped (**ADR-0021**) → bot live 2026-09-20. Live status in **#106**. Draft: **#109**.

**Comment by bgauduch** [comment]
Pointer for whoever implements this: closed PR **#109** contains a working draft of the Renovate **custom regex managers** for the tool versions in `supported_versions.json` (`hashicorp/terraform` via `github-releases`, `aws/aws-cli` via `github-tags`).

The general Renovate config already shipped in **#122**; only this custom-manager part remains and is tracked here. Salvage #109's regex managers as the starting point.

---
_Generated by [Claude Code](https://claude.ai/code)_

**Comment by bgauduch** [comment]
Arbitration proposed in #152 (harness study, ADR-0014 phase 2): this issue is **not** an alternative to a bump script, it is its trigger.

- **Renovate customManager** on `supported_versions.json` does the *detection* and opens the PR.
- A deterministic `scripts/bump-version.sh` does the *materialisation* that Renovate cannot: fetch `SHA256SUMS` and `.sig` into `security/**`, GPG verify them, drop the superseded patch of the same minor per ADR-0004, and sync `tests/container-structure-tests.yml.template`.
- A new fast validation oracle (`scripts/validate.sh --fast`, #152 PR 1) proves the result before push: `supported_versions.json` against `security/**`, and the version policy itself.

That split is what unblocks the current deferral. Sequenced as #152 PR 4, after the oracle lands.

**Comment by bgauduch** [comment]
## Evidence that rung 0 of the automation ladder is not running at all

Measured on 2026-08-23, on `master`. This issue owns the custom manager and the dependency-coverage audit; what follows is the coverage audit's first result, and it is more basic than a missing custom manager.

**Renovate has produced nothing since it was configured.** `renovate.json` landed in #122 (Phase 1, 2026-06) and retired Dependabot in the same PR. Since then:

- no commit on `master` matches `chore(deps)` or `renovate` — the last one is #122 itself, which is the config;
- no open pull request is Renovate's (`list_pull_requests` returns only the community fork #88);
- no branch is Renovate's (`master` plus session branches only);
- **no Dependency Dashboard issue exists**, although `renovate.json` extends `:dependencyDashboard`. Renovate opens that issue on its first run against a repository.

The last point is the decisive one: the dashboard's absence is evidence the bot has never run here, not that it ran and found nothing. The most likely cause is that the Renovate GitHub App was never installed on the repository — the config is a config, not an installation.

**What that cost, in things this repository has already paid for by hand:**

| Drift | State on 2026-08-23 | Renovate's rung-0 job |
|---|---|---|
| Debian base digest | pinned `sha256:020c0d20…`; `trixie-slim` upstream is `sha256:3a39a059…`, published 2026-08-05 | `dockerfile` manager, grouped as `debian-base-image` |
| GitHub Actions | not audited here | `github-actions` manager, grouped |
| APT pins | `jq`/`unzip` pins superseded, `master` unbuildable until the hand-written bump in #168 | not Renovate's — rung 1, `refresh-apt-pins.sh` (#152 PR 4) |
| AWS CLI | `2.36.6`; upstream is `2.36.29` | the custom manager this issue is about |
| Terraform | `1.15.8`; upstream is `1.15.9`, and `1.16.0-rc2` is out | same |

**Consequence for #152.** Its automation ladder lists "GitHub Actions, Debian base digest → rung 0, bot, Renovate → **done**". That row is aspirational: nothing is automated at rung 0 today, so every rung above it is being reached by hand. Verifying the installation is a precondition of this issue, not a detail of it — a custom manager added to a config no bot reads changes nothing.

**Second question the install raises**, worth settling in the same pass: `renovate.json` sets `automerge: true` for `patch`/`pin`/`digest`. If branch protection requires a review, that setting is inert; if it does not, a digest bump of the base image reaches `master` — and `push-edge` republishes `edge` from it — with no human in the loop. ADR-0012 gives the human the merge; whether a bot is in scope of that rule has never been decided.

Suggested next step, maintainer-side and outside any session's reach: confirm whether the Renovate App is installed on `bgauduch/terraform-aws-cli` (https://github.com/apps/renovate). Everything else in this issue is downstream of that answer.

---
_Generated by [Claude Code](https://claude.ai/code)_

**Comment by bgauduch** [comment]
Folded into the body (dated note in the header + Task 4, 2026-09-20).

---
_Generated by [Claude Code](https://claude.ai/code)_
